src/main/clojure/clara/rules/compiler.clj

```diff
@@ -235,15 +235,21 @@
   "Compiles a sequence of constraints into a structure that can be evaluated.
 
    Callers may also pass a collection of equality-only-variables, which instructs
-   this function to only do an equality check on them rather than create a unification binding."
+   this function to only do an equality check on them rather than create a unification binding.
+
+   The returned expression evaluates to a two item vector, [bindings error], which contains
+   the evaluated bindings in the first argument and a map with information on any exception
+   that might have been thrown by evaluating a constraint as the second vector entry."
```

Comment:
Looking at this again I think I may have misinterpreted parts of @rbrush 's comment.  We could do something like

````
;;; compiler
(compile-constraints
 (try constraint-stuff
      (catch Exception e
        (throw (ex-info "message" {:exception e
                                   :constraint-exp val-here
                                   :bindings val-here
                                   :env env
                                   ::constraint-exception true})))))
;; engine
(defn- alpha-node-matches
  [facts env activation node]
  (for [fact facts
        :let [bindings (try (activation fact env)
                            (catch #?(:clj Exception :cljs :default) e
                              (if (-> e ex-data :clara.rules.compiler/constraint-exception)
                                (throw-constraint-exception
                                 (-> e ex-data
                                     (dissoc :clara.rules.compiler/constraint-exception)
                                     node
                                     fact
                                     env))
                                (throw e)
                                ;; Or add some info in the else if we want but that isn't necessary for this enhancement.
                                )))]
        :when bindings] ; FIXME: add env.
    [fact bindings]))
````

I think this is what Ryan was actually getting at, but he can speak to whether my first interpretation or this one is correct, or if I was misinterpreting both times. :)  This sort of approach is probably even easier than what I suggested before.  The one thing I did change here is to use a key in an ex-data map to dispatch on whether the exception was thrown from a constraint rather than using a different exception type which I think is probably easier than needing to define a new exception type for both Clojure and ClojureScript.